processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,385 @@
+
+  private void initNewSlot(ByteBuffer newKey)
+  {
+    increaseWriteIndex();
+
+    final int recordOffset = recordSize * curWriteIndex;
+    buffer.position(recordOffset);
+    buffer.put(newKey);
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(buffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  /**
+   * Wait for {@link #nextReadIndex} to be moved if necessary and move {@link #curWriteIndex}.  {@link #nextReadIndex}
+   * is checked in while loops instead of waiting using a lock to avoid frequent thread park.
+   */
+  private void increaseWriteIndex()
+  {
+    if (curWriteIndex == maxSlotNum - 1) {
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextReadIndex == -1 || nextReadIndex == 0) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextWriteIndex == nextReadIndex) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+      }
+      curWriteIndex = nextWriteIndex;
+    }
+  }
+
+  @Override
+  public void reset()
+  {
+    curWriteIndex = -1;
+    nextReadIndex = -1;
+    finished = false;
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  /**
+   * Signal that no more inputs are added.  Must be called after {@link #aggregate(Object)} is called for the last input.
+   */
+  public void finish()
+  {
+    increaseWriteIndex();
+    finished = true;
+  }
+
+  /**
+   * Return a sorted iterator.  This method can be called safely while writing and iterating thread and writing thread
+   * can be different.  The result iterator always returns sorted results.  This method should be called only one time
+   * per grouper.
+   *
+   * @return a sorted iterator
+   */
+  public CloseableIterator<Entry<KeyType>> iterator()
+  {
+    if (!initialized) {
+      throw new ISE("Grouper should be initialized first");
+    }
+
+    return new CloseableIterator<Entry<KeyType>>()
+    {
+      {
+        // Waits for some data to be ready
+        final long startLoopAt = System.currentTimeMillis();
+        while ((curWriteIndex == -1 || curWriteIndex == 0) && !finished && !Thread.interrupted()) {
+          if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+            throw new RuntimeException(new TimeoutException());
+          }
+        }
+        nextReadIndex = 0;
+      }
+
+      @Override
+      public boolean hasNext()
```

Comment:
It may return true, then false, then true again, doesn't conform the Iterator interface

Reply:
How could it be possible? If it returns false for the first time, then ```finished``` should be true and remaining() should return 0. Once ```finished``` becomes true, ```curWriteIndex``` is not changed, and thus ```remaining()``` should be always reduced as ```nextReadIndex``` increases.